util/env_windows.cc

```diff
@@ -0,0 +1,920 @@
+
+  // REQUIRES: mu_ must be held
+  void SetAllowed(intptr_t v) {
+    allowed_.Release_Store(reinterpret_cast<void*>(v));
+  }
+
+  Limiter(const Limiter&);
+  void operator=(const Limiter&);
+};
+
+
+class WindowsSequentialFile: public SequentialFile {
+ private:
+  std::string filename_;
+  HANDLE handle_;
+
+ public:
+  WindowsSequentialFile(const std::string& fname, HANDLE handle)
+      : filename_(fname), handle_(handle) {}
+  virtual ~WindowsSequentialFile() { CloseHandle(handle_); }
+
+  virtual Status Read(size_t n, Slice* result, char* scratch) {
+    DWORD r = 0;
+    if (!::ReadFile(handle_, scratch, n, &r, NULL)) {
+        return WindowsError(filename_, GetLastError());
+    }
+    *result = Slice(scratch, r);
+    return Status::OK();
+  }
+
+  virtual Status Skip(uint64_t n) {
+    LARGE_INTEGER bytes;
+    bytes.QuadPart = n;
+    if (!::SetFilePointerEx(handle_, bytes, NULL, FILE_BEGIN)) {
+        return WindowsError(filename_, GetLastError());
+    }
+    return Status::OK();
+  }
+};
+
+
+class WindowsRandomAccessFile: public RandomAccessFile {
+ private:
+  std::string filename_;
+  bool temporary_handle_;  // If true, we open on every read.
+  HANDLE handle_;
+  Limiter* limiter_;
+
+ public:
+  WindowsRandomAccessFile(const std::string& fname, HANDLE handle, Limiter* limiter)
+      : filename_(fname), handle_(handle), limiter_(limiter) {
+    temporary_handle_ = !limiter->Acquire();
+    if (temporary_handle_) {
+      // Open file on every access.
+      CloseHandle(handle_);
+      handle_ = INVALID_HANDLE_VALUE;
+    }
+  }
+
+  virtual ~WindowsRandomAccessFile() {
+    if (!temporary_handle_) {
+      CloseHandle(handle_);
+      limiter_->Release();
+    }
+  }
+
+  virtual Status Read(uint64_t offset, size_t n, Slice* result,
+                      char* scratch) const {
+    HANDLE handle = handle_;
+    if (temporary_handle_) {
+      // Create temporary handle.
+      handle = CreateFileW(
+        Utf8ToUtf16(filename_).data(),
+        GENERIC_READ,
+        FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+        NULL,
+        OPEN_EXISTING,
+        FILE_FLAG_RANDOM_ACCESS,
+        NULL
+      );
+      if (handle == INVALID_HANDLE_VALUE) {
+        return WindowsError(filename_, GetLastError());
+      }
+    }
+
+    pread_mutex.Lock();
+
+    // Seek offset
+    Status s = Status::OK();
+    LARGE_INTEGER in, out;
+    in.QuadPart = offset;
+    if (!SetFilePointerEx(handle, in, &out, FILE_BEGIN)) {
+      s = WindowsError(filename_, GetLastError());
+    }
+
+    // read from file
+    DWORD r = 0;
+    if (s.ok() && !::ReadFile(handle, scratch, n, &r, NULL)) {
+      s = WindowsError(filename_, GetLastError());
+    }
+
+    pread_mutex.Unlock();
+
+    *result = Slice(scratch, r);
+    if (temporary_handle_) {
+      // Close the temporary file descriptor opened earlier.
+      CloseHandle(handle);
+    }
+    return s;
+  }
+};
+
+
+class WindowsMmapReadableFile: public RandomAccessFile
```

Comment:
I think we should leave out the mmap support in this initial PR.  Consider a separate PR for that later on backed by real performance data as to how much mmap based reading helps.